### PR TITLE
chore: add clock pkg for testing time

### PR DESCRIPTION
**Pull request by spikecurtis**
Adds a package for testing time/timer/ticker functions.  Implementation is limited to `NewTimer` and `NewContextTicker`, but will eventually be expanded to all `time` functions from the standard library as well as `context.WithTimeout()`, `context.WithDeadline()`.

Replaces `benbjohnson/clock` for the pubsub watchdog, as a proof of concept.

Eventually, as we expand functionality, we will replace most time-related functions with this library for testing.

### CODE REVIEW

**Comment by spikecurtis** [comment]
* **#13461** <a href="https://app.graphite.dev/github/pr/coder/coder/13461?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: clock/clock.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"context"
+	"time"
+)
+
+type Clock interface {
+	// NewContextTicker is a convenience function that calls f on the interval d until either the
+	// given context expires or f returns an error.  Callers may call Wait() on the returned Waiter
+	// to wait until this happens and obtain the errror.
+	NewContextTicker(ctx context.Context, d time.Duration, f func() error, tags ...string) Waiter
```
Comment (mafredri) [inline clock/clock.go:16]: How would you feel about using functional options instead of `...string`? An alternative to having this convenience function would be to accept the following signature:

`NewTicker(d time.Duration, f func() error, opts ...Option)` and you'd pass it `clock.WithContext(ctx)` or `clock.Tags("tag1", "tag2") as options.

I think this could result in a slightly cleaner and more flexible API, as well as the added verbosity/explicitness can help with code reads (i.e. "what are these strings?").

In this scenario, I imagine the return signature of `NewTicker` could include all three methods `Stop`, `Reset` and _`Wait`_.

File: coderd/database/pubsub/watchdog_test.go [hunk 2]
```diff
@@ -4,36 +4,48 @@ import (
...
-	start := time.Date(2024, 2, 5, 8, 7, 6, 5, time.UTC)
-	mClock.Set(start)
 	logger := slogtest.Make(t, &slogtest.Options{IgnoreErrors: true}).Leveled(slog.LevelDebug)
 	fPS := newFakePubsub()
+
+	// trap the ticker and timer.Stop() calls
+	pTrap := mClock.Trap().NewContextTicker("publish")
+	defer pTrap.Close()
+	sTrap := mClock.Trap().TimerStop("subscribe")
+	defer sTrap.Close()
```
Comment (mafredri) [inline coderd/database/pubsub/watchdog_test.go:27]: Using slightly longer names could be beneficial here so that it's easier to read. The one letter difference between traps is a bit hard to follow after a certain distance.

File: coderd/database/pubsub/watchdog_test.go [hunk 3]
```diff
@@ -4,36 +4,48 @@ import (
...
+
 	uut := pubsub.NewWatchdogWithClock(ctx, logger, fPS, mClock)
 
+	// wait for the ticker to be created so that we know it starts from the
+	// right baseline time.
+	pc, err := pTrap.Wait(ctx)
+	require.NoError(t, err)
+	require.Equal(t, 15*time.Second, pc.Duration)
+	pc.Release()
```
Comment (mafredri) [inline coderd/database/pubsub/watchdog_test.go:36]: The require(s) above may result in release never being called. Just pointing this out in case it matters (e.g. will we leave goroutines behind, block in some other defer, etc?).

**Review by mafredri — COMMENTED** [review]
Nice work! I'm happy to see how the modified test cases in pubsub nicely demonstrates the utility of the trap concept. I didn't re-read the RFC but this implementation largely matches my recollection of the approach.

Considering this is likely going to be its own library in the future, I'd love to see some `func Example...`s that demonstrate the functionality and use-cases. Some in-package tests would be nice too.

File: clock/clock.go (hunk 1 above)
Comment (spikecurtis) [reply clock/clock.go:16]: Firstly, maybe "ContextTicker" isn't the best name.  It's not a ticker that optionally takes a context, it's a ticker calls the function `f` on the ticks, rather than just sending the ticks to a channel.  Sort of like `AfterFunc` is a special timer that calls a function.  "TickerFunc"?

Secondly, the idea is to include tags on _every_ call in the `clock` interface, and all the timer/ticker calls as well.  With the exception of this ContextTicker/TickerFunc, the idea is to verbatim replicate the functionality of the `time` package, not enhance it. So, I literally cannot think of another option we'd want other than tags, and even if I could, it's unlikely to apply to _every_ call, so adding it to a generic `Option` interface would muddy the waters. I think variability is actually a downside in this case, and we should keep it simple. 

Lastly, I'm not a fan of the verbosity of `clock.Tags("tag1")`.  The signature says `tags ...string`, so I feel like that's clear enough once you jump to the definition.

File: clock/clock.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	NewTimer(d time.Duration, tags ...string) *Timer
```
Comment (johnstcn) [inline clock/clock.go:17]: Missing documentation for `NewTimer`.

File: clock/mock.go [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+
+func (t Trapper) ContextTickerWait(tags ...string) *Trap {
+	return t.mock.newTrap(clockFunctionContextTickerWait, tags)
+}
+
+func (m *Mock) Trap() Trapper {
+	return Trapper{m}
+}
+
+func (m *Mock) newTrap(fn clockFunction, tags []string) *Trap {
```
Comment (johnstcn) [inline clock/mock.go:247]: Is it possible to add a trap for a tag that does not exist? If so, can this be caught early and immediately call `t.Fatal()`?

File: clock/mock.go
Comment (johnstcn) [inline clock/mock.go:1]: It would be useful to add an example test for this package.

**Comment by johnstcn** [comment]
I really like how this is looking so far! 👍 Just a few suggestions on my end.

File: clock/mock.go [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+	"context"
+	"errors"
+	"slices"
+	"sync"
+	"time"
+)
+
+// Mock is the testing implementation of Clock.  It tracks a time that monotonically increases
+// during a test, triggering any timers or tickers automatically.
+type Mock struct {
```
[the remark this answers is not in the document]
Comment (spikecurtis) [reply clock/mock.go:13]: It's hard to split the mock implementation from the real implementation if we want to keep `Timer` as a `struct` with a `C` field, like it is in `time`.  That minimizes code changes in moving from the `time` package to this.

How we do it today is to have a single `struct` that includes both the real and mock implementations.  If `Timer` were an interface we could split them, but then you'd have to change `<-tmr.C` to `<-tmr.C()` everywhere.

File: clock/mock.go
Comment (spikecurtis) [reply clock/mock.go:1]: Up the stack

**Review by johnstcn — APPROVED** [review]
